### ISSUE TITLE
Mutli-role functionality conflicts if secondary role is higher than customer

### ISSUE THREAD

**Issue by krogsgard** [woothemes/woocommerce#10612]
I ran into this today in WC 2.5.5.

I have a customer who I promoted to an Editor, so the role went from Customer to Editor.

The customer then bought something (in this case, we let a renewal of a subscription go through), and it added a secondary Customer role (I guess something newish in WC?).

The customer -- acting as Editor, which obviously WP can't tell between how they're acting :) -- goes back to log into the admin, but is blocked from viewing the admin because they are a "Customer" role or some cap check (sorry, I haven't looked to see the admin viewage block).

So even though they have the Editor role, bc of the secondary Customer role now, they can't see the admin.

So I go change their role on the admin profile from "Editor" to Editor and press Save. Now it goes back to one role, Editor, and they can view it again.

I have no idea how common this is or if maybe I have some other juju going on. Partial Status Report going below but redacted a lot of status info (ask me privately if you want)

```
### WordPress Environment ###

WC Version: 2.5.5
Log Directory Writable: ✔
WP Version: 4.4.2
WP Multisite: –
WP Memory Limit: 512 MB
WP Debug Mode: –
Language: en_US

### Server Environment ###

PHP Version: 5.5.30-1+deb.sury.org~precise+1
MySQL Version: 5.6.10

Subscriptions Mode: ✔ 
						
							 Live

### Database ###

WC Database Version: 2.5.5

### API ###

API Enabled: ✔
API Version: 3.1.0

```

**Comment by BurlesonBrad** [comment]
I think I need one :beer: to read this, 2 :beers: to understand it, and a :cocktail: to answer.

_ok, follow me here, and see if I make any sense_

1. The first assignment of editor **did** work, but perhaps they never logged in between that date and the date that the subscription was renewed.
2. Subscription renews, which triggers role to go back to Customer.
3. You manually click editor again. This time they **immediately** check to make sure they can log in, and ~of course~ they can.
4. as of this 'issue' date, their subscription has **not** renewed _**(yet)**_ has it?

**Comment by krogsgard** [comment]
You're close, but not quite there... I'll use screenshots :smile: 

Step 0: I made someone that was already a customer some other role (author, editor). Note: it does _not_ apply to Shop Manager (that works as expected).

Here's an instance where I previously made some folks authors:

https://cloudup.com/csYmduig4jo

Step 1: Some of these folks with the "author" role previously have had their subs renew or purchased a product, so they got the "customer" role back, but bc of new stuff in WC it's not a dual-role, not _just_ Customer or _just_ Author. They are a customer and an author.

Step 2: Switch to them (or they log in and try to be an author). An it won't let them see the admin, because they are a Customer, even though an Author should be able to see the admin.

https://cloudup.com/cE_AM0hqozg

Step 3: Go in as an administrator and update that user on the "personal options" page. You don't have to change anything, it's already set in the dropdown as "Author" because there is no multi-select for multiple roles, even though WC wizards that in.

https://cloudup.com/cs7lnUcfHpg

Step 4: Go back to `wp-admin/users.php?role=author` (or whatever non Shop Manager traditional WP role), and they have changed to one role again, only "Author" or "Editor" and no longer includes "Customer" -- and they can now view the admin again.

What it means:

Basically, the product order makes them a dual role, and WC isn't checking for the proper scope of caps users may already have for admin privileges (should be everything but what a Subscriber has, probably), and therefore limits their admin access where it should not, until you re-save their profile page, but then it (properly improperly) removes the handy dual-role Customer flag.

Make better sense now?

**Comment by javorszky** [comment]
Subscriptions uses `add_role` and `remove_role` instead of `set_role` because if shops have other user data associated with the user (for example role based permission / access rules), then a single `set_role` would undo all of that.

For those who can access the Prospress repo: https://github.com/Prospress/woocommerce-subscriptions/issues/922

**Comment by javorszky** [comment]
From WC Core: https://github.com/woothemes/woocommerce/blob/master/includes/admin/class-wc-admin.php#L116-L132

An editor by itself has the cap `edit_posts`, but when a `customer` role is also added to that user, the `edit_posts` cap will become false. This piece of code then decides that admin access is a no-go: https://github.com/woothemes/woocommerce/blob/master/includes/admin/class-wc-admin.php#L116-L132

* Only editor: https://cloudup.com/c5-ijMMxagH
* Editor + customer: https://cloudup.com/c6_svGQRaSq

Hope that helps. No idea at this point whether this is a WP core thing (`map_meta_cap`) or a WC core thing.

**Comment by javorszky** [comment]
Fun fact, when `subscriber` role is added to the `editor` role, the `edit_posts` cap doesn't change.

**Comment by BurlesonBrad** [comment]
#learning_by_contributing_on_github 
@krogsgard: made sense :+1:  & @javorszky now sheds light :flashlight: 
lemme quench my gut instinct & throw a mention over to @claudiosmweb. Something tells me he'd be able to come up with a creative solution.

**Comment by mikejolley** [comment]
So, WooCommerce itself does not set, add, or remove roles from users. Users are left as-is. The only time a customer role is used is on user creation. Thats not the case here.

I'll look into the permission checks on dual role users - this may be a WP core bug in which case I'll try squashing it.

**Comment by BurlesonBrad** [comment]
@mikejolley isn't the _"fix"_ within Subscriptions?

**Comment by mikejolley** [comment]
@BandonRandon maybe not.

Replicated with https://dl.dropboxusercontent.com/s/xgucqvvh6no3skm/2016-03-30%20at%2017.49.png?dl=0

`current_user_can( 'edit_posts' )` is false. So this isn't a WooCommerce issue, but looks like a WordPress bug.

@krogsgard I'll close this out since it's not us but I'm working on a core fix (scouting for already reported issues on trac first).

**Comment by mikejolley** [comment]
@krogsgard https://core.trac.wordpress.org/ticket/36376#comment:1

**Comment by BandonRandon** [comment]
I think you mean @BurlesonBrad, Mike :)
